File: sshd/src/main/java/jetbrains/exodus/sshd/RhinoCommand.java
/**
 * Copyright 2010 - 2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.sshd;

import jetbrains.exodus.entitystore.PersistentEntityStore;
import jetbrains.exodus.entitystore.StoreTransaction;
import jetbrains.exodus.entitystore.StoreTransactionalExecutable;
import org.apache.sshd.server.Command;
import org.apache.sshd.server.Environment;
import org.apache.sshd.server.ExitCallback;
import org.jetbrains.annotations.NotNull;
import org.mozilla.javascript.*;
import org.mozilla.javascript.tools.ToolErrorReporter;

import java.awt.event.KeyEvent;
import java.io.*;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 *
 */
public class RhinoCommand implements Command, Runnable {

    private static final String[] INITIAL_SCRIPTS = {"init.js", "functions.js", "opt.js"};
    private InputStreamReader in;
    private PrintStream out;
    private PrintStream err;
    private ExitCallback callback;
    private Scriptable scope;
    private volatile boolean stopped = false;
    private final Object lock = new Object();

    @NotNull
    private Map<String, Object> config;

    RhinoCommand(@NotNull Map<String, Object> config) {
        this.config = config;
    }

    @Override
    public void setInputStream(InputStream in) {
        try {
            this.in = new InputStreamReader(in, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void setOutputStream(OutputStream out) {
        this.out = new PrintStream(out);
    }

    @Override
    public void setErrorStream(OutputStream err) {
        this.err = new PrintStream(err);
    }

    @Override
    public void setExitCallback(ExitCallback callback) {
        this.callback = callback;
    }

    @Override
    public void start(Environment env) throws IOException {
        new Thread(this).start();
    }

    @Override
    public void destroy() {
        stopped = true;
    }

    @Override
    public void run() {
        Context cx = Context.enter();
        cx.setClassShutter(new ClassShutterImpl());
        cx.setErrorReporter(new ErrorReporterImpl());
        try {
            processInput(cx);
        } finally {
            callback.onExit(0);
            synchronized (lock) {
                try {
                    if (scope != null) {
                        evalResourceScript(cx, scope, "destroy.js");
                        scope = null;
                    }
                } finally {
                    stopped = true;
                    Context.exit();
                }
            }
        }
    }

    private String readLine(API api) {
        StringBuilder s = new StringBuilder();

        try {
            while (true) {
                char c = (char) in.read();

                if (c == -1 || c == 3) return null;
                if (c == '\n' || c == '\r') {
                    api.newLine();
                    break;
                }
                // delete
                if (c == 127 && s.length() > 0) {
                    s.deleteCharAt(s.length() - 1);
                    api.printChar('\b');
                }

                if (!isPrintableChar(c)) continue;

                api.printChar(c);
                s.append(c);
            }
            return s.toString();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private boolean isPrintableChar(char c) {
        Character.UnicodeBlock block = Character.UnicodeBlock.of(c);
        return (!Character.isISOControl(c)) &&
                c != KeyEvent.CHAR_UNDEFINED &&
                block != null &&
                block != Character.UnicodeBlock.SPECIALS;
    }

    private Scriptable createScope(Context cx, API api) {
        final Scriptable scope = cx.initStandardObjects(null, true);
        NativeObject config = new NativeObject();
        for (Map.Entry<String, Object> entry : this.config.entrySet()) {
            config.defineProperty(entry.getKey(), entry.getValue(), NativeObject.READONLY);
        }
        config.defineProperty("implementationVersion", cx.getImplementationVersion(), NativeObject.READONLY);
        ScriptableObject.putProperty(scope, "config", config);
        ScriptableObject.putProperty(scope, "api", Context.javaToJS(api, scope));
        for (String name : INITIAL_SCRIPTS) {
            evalResourceScript(cx, scope, name);
        }
        return scope;
    }

    private void evalResourceScript(Context cx, Scriptable scope, String name) {
        try {
            cx.evaluateReader(scope, new InputStreamReader(this.getClass().getResourceAsStream(name)), name, 1, null);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void processInput(final Context cx) {
        while (true) {
            final API api = new API();

            if (scope == null) {
                scope = createScope(cx, api);
            }

            out.flush();
            int lineno = 0;
            final StringBuilder source = new StringBuilder();

            // Collect lines of source to compile.
            while (true) {
                lineno++;
                api.printChar(lineno == 1 ? '>' : ' ');
                String newline = readLine(api);
                if (newline == null) return;
                source.append(newline).append("\n");
                if (cx.stringIsCompilableUnit(source.toString())) break;
            }

            // execute script in transaction
            try {
                final Script script = cx.compileString(source.toString(), "<stdin>", lineno, null);
                if (script != null) {
                    long start = System.currentTimeMillis();

                    PersistentEntityStore entityStore = getPersistentStore(scope);
                    if (entityStore == null) {
                        processScript(api, script, cx, scope);
                    } else {
                        entityStore.executeInTransaction(new StoreTransactionalExecutable() {
                            @Override
                            public void execute(@NotNull StoreTransaction txn) {
                                ScriptableObject.putProperty(scope, "txn", Context.javaToJS(txn, scope));
                                processScript(api, script, cx, scope);
                            }
                        });
                    }

                    api.println("Complete in " + ((System.currentTimeMillis() - start)) + " ms");
                }
            } catch (RhinoException rex) {
                ToolErrorReporter.reportException(cx.getErrorReporter(), rex);
            } catch (VirtualMachineError ex) {
                // Treat StackOverflow and OutOfMemory as runtime errors
                ex.printStackTrace();
                Context.reportError(ex.toString());
            }

            if (api.reset) {
                synchronized (lock) {
                    evalResourceScript(cx, scope, "destroy.js");
                    scope = null;
                }
            }
        }
    }

    private PersistentEntityStore getPersistentStore(Scriptable scope) {
        Object store = scope.get("store", null);
        if (store == null || store == UniqueTag.NOT_FOUND) {
            return null;
        }
        return (PersistentEntityStore) Context.jsToJava(store, PersistentEntityStore.class);
    }

    protected void processScript(API api, Script script, Context cx, Scriptable scope) {
        Object result = script.exec(cx, scope);
        if (result != Context.getUndefinedValue()) {
            api.println(Context.toString(result));
        }
    }

    public class API {
        private boolean reset;

        public void reset() {
            reset = true;
        }

        @Override
        public String toString() {
            return "API";
        }

        public void print(String s) {
            out.print(s == null ? null : s.replace("\n", "\n\r"));
            out.flush();
        }

        public void println(String s) {
            print(s + "\n\r");
        }

        public void newLine() {
            out.print("\n\r");
            out.flush();
        }

        public void printChar(char s) {
            if (s == '\n') {
                newLine();
            } else {
                out.print(s);
                out.flush();
            }
        }
    }

    private class ErrorReporterImpl implements ErrorReporter {
        @Override
        public void error(String message, String sourceName, int line, String lineSource, int lineOffset) {
            StringBuilder s = new StringBuilder();
            if (line > 0) {
                s.append(line).append(": ");
            }
            s.append(message);
            err.print(s.toString() + "\n\r");
            err.flush();
        }

        @Override
        public EvaluatorException runtimeError(String message, String sourceName, int line, String lineSource, int lineOffset) {
            return new EvaluatorException(message, sourceName, line, lineSource, lineOffset);
        }

        @Override
        public void warning(String message, String sourceName, int line, String lineSource, int lineOffset) {
        }
    }

    private static Set<String> FORBIDDEN_CLASSES = new HashSet<String>() {{
        add("java.lang.System");
    }};

    private class ClassShutterImpl implements ClassShutter {

        @Override
        public boolean visibleToScripts(String fullClassName) {
            return !FORBIDDEN_CLASSES.contains(fullClassName);
        }
    }
}

File: sshd/src/main/kotlin/jetbrains/exodus/javascript/RhinoCommand.kt
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.javascript

import jetbrains.exodus.core.dataStructures.NanoSet
import jetbrains.exodus.core.dataStructures.Priority
import jetbrains.exodus.core.execution.Job
import jetbrains.exodus.entitystore.PersistentEntityStoreImpl
import jetbrains.exodus.env.EnvironmentImpl
import jetbrains.exodus.kotlin.notNull
import org.apache.sshd.server.Command
import org.apache.sshd.server.Environment
import org.apache.sshd.server.ExitCallback
import org.mozilla.javascript.*
import org.mozilla.javascript.tools.ToolErrorReporter
import java.awt.event.KeyEvent
import java.io.*

abstract class RhinoCommand(protected val config: Map<String, *>) : Job(), Command {

    companion object {

        internal fun createCommand(config: Map<String, *>): RhinoCommand {
            return when (config[API_LAYER]) {
                ENVIRONMENTS -> EnvironmentRhinoCommand(config)
                ENTITY_STORES -> EntityStoreRhinoCommand(config)
                else -> throw IllegalArgumentException("The value of apiLayer should be 'Environments' or 'EntityStores'")
            }
        }

        const val TXN_PARAM = "txn"
        const val API_LAYER = "apiLayer"
        const val ENVIRONMENTS = "environments"
        const val ENTITY_STORES = "entitystores"
        const val CONSOLE = "console"
        const val ENVIRONMENT_INSTANCE = "environment instance"
        const val ENTITY_STORE_INSTANCE = "entity store instance"
        private const val JS_ENGINE_VERSION_PARAM = "jsEngineVersion"
        private const val CONFIG_PARAM = "config"
        private const val INTEROP_PARAM = "interop"
        private val FORBIDDEN_CLASSES = NanoSet("java.lang.System")
        private const val RECENT_COMMANDS_LIST_MAX_SIZE = 1000

        private fun isPrintableChar(c: Char): Boolean {
            val block = Character.UnicodeBlock.of(c)
            return !Character.isISOControl(c) &&
                    c != KeyEvent.CHAR_UNDEFINED &&
                    block != null &&
                    block !== Character.UnicodeBlock.SPECIALS
        }
    }

    private var callback: ExitCallback? = null
    private var input: InputStream? = null
    private var stopped = false
    private var output: PrintStream? = null
    private var error: PrintStream? = null
    private val recentCommands: MutableList<String> = arrayListOf()

    override fun setExitCallback(callback: ExitCallback?) {
        this.callback = callback
    }

    override fun setInputStream(input: InputStream?) {
        this.input = input
    }

    override fun destroy() {
        stopped = true
    }

    override fun setErrorStream(err: OutputStream?) {
        error = toPrintStream(err)
    }

    override fun setOutputStream(out: OutputStream?) {
        output = toPrintStream(out)
    }

    override fun start(env: Environment?) {
        processor = RhinoServer.commandProcessor
        queue(Priority.normal)
    }

    override fun execute() {
        val cx = Context.enter()
        cx.setClassShutter { !FORBIDDEN_CLASSES.contains(it) }
        cx.errorReporter = object : ErrorReporter {
            override fun warning(message: String?, sourceName: String?, line: Int, lineSource: String?, lineOffset: Int) {
                // ignore
            }

            override fun runtimeError(message: String?, sourceName: String?, line: Int, lineSource: String?, lineOffset: Int): EvaluatorException {
                return EvaluatorException(message, sourceName, line, lineSource, lineOffset)
            }

            override fun error(message: String?, sourceName: String?, line: Int, lineSource: String?, lineOffset: Int) {
                val out = output.notNull
                if (sourceName != null) {
                    out.print("$sourceName: ")
                }
                if (line > 0) {
                    out.print("$line: ")
                }
                out.println(message)
                out.flush()
            }
        }
        try {
            processInput(cx)
        } finally {
            try {
                callback?.onExit(0)
            } finally {
                stopped = true
                Context.exit()
            }
        }
    }

    protected fun evalResourceScripts(cx: Context, scope: Scriptable, vararg names: String) {
        names.forEach {
            cx.evaluateReader(scope, InputStreamReader(this.javaClass.getResourceAsStream(it)), it, 1, null)
        }
    }

    fun evalFileSystemScript(cx: Context, scope: Scriptable, fileName: String) {
        cx.evaluateReader(scope, FileReader(fileName), fileName, 1, null)
    }

    protected fun processScript(interop: Interop, script: Script, cx: Context, scope: Scriptable?) {
        val result = script.exec(cx, scope)
        if (result !== Context.getUndefinedValue()) {
            interop.println(result)
        }
    }

    protected open fun evalInitScripts(cx: Context, scope: Scriptable) {
        evalResourceScripts(cx, scope, "bindings.js", "functions.js")
    }

    protected abstract fun evalTransactionalScript(cx: Context, script: Script, interop: Interop, scope: Scriptable)

    private fun processInput(cx: Context) {
        Interop(this, output.notNull,
                config[ENVIRONMENT_INSTANCE] as? EnvironmentImpl,
                config[ENTITY_STORE_INSTANCE] as? PersistentEntityStoreImpl).use { interop ->
            interop.flushOutput()
            val scope = createScope(cx, interop)
            evalInitScripts(cx, scope)
            while (!stopped) {
                interop.printPrompt()
                val cmd = buildString { readLine(this, interop) }
                try {
                    if (cmd.isNotBlank()) {
                        // exit
                        if (cmd.equals("exit", true) || cmd.equals("quit", true)) {
                            if (isConsole) {
                                interop.println("Press Enter to $cmd...")
                            }
                            break
                        }
                        recentCommands.add(cmd)
                        if (recentCommands.size > RECENT_COMMANDS_LIST_MAX_SIZE) {
                            recentCommands.removeAt(0)
                        }
                        val script = cx.compileString(ScriptPreprocessor.preprocess { cmd }, "<stdin>", 0, null)
                        if (script != null) {
                            val start = System.currentTimeMillis()
                            // execute script in transaction if an Environment or an EntityStore is open
                            evalTransactionalScript(cx, script, interop, scope)
                            interop.println("Completed in ${(System.currentTimeMillis() - start)} ms")
                        }
                    }
                } catch (rex: RhinoException) {
                    ToolErrorReporter.reportException(cx.errorReporter, rex)

                } catch (ex: VirtualMachineError) {
                    // Treat StackOverflow and OutOfMemory as runtime errors
                    ex.printStackTrace()
                    Context.reportError(ex.toString())
                }
            }
        }
    }

    private fun createScope(cx: Context, interop: Interop): Scriptable {
        interop.cx = cx
        val scope = cx.initStandardObjects(null, true)
        interop.scope = scope
        val config = NativeObject()
        this.config.forEach {
            config.defineProperty(it.key, it.value, NativeObject.READONLY)
        }
        config.defineProperty(JS_ENGINE_VERSION_PARAM, cx.implementationVersion, NativeObject.READONLY)
        ScriptableObject.putProperty(scope, CONFIG_PARAM, config)
        ScriptableObject.putProperty(scope, INTEROP_PARAM, Context.javaToJS(interop, scope))
        return scope
    }

    private fun readLine(s: StringBuilder, interop: Interop) {
        val inp = input ?: return
        var cmdIdx = recentCommands.size
        while (!stopped) {
            val c = inp.read()
            if (c == -1 || c == 3 || c == 26) break
            val ch = c.toChar()
            if (ch == '\n') {
                if (!isConsole) {
                    interop.newLine()
                }
                break
            }
            when (c) {
            // backspace
                127 ->
                    if (s.isNotEmpty()) {
                        s.deleteCharAt(s.length - 1)
                        interop.backspace()
                    }
            // up or down
                27 -> {
                    inp.read()
                    var newCmdIdx = cmdIdx
                    val upOrDown = inp.read()
                    if (upOrDown == 65) {
                        --newCmdIdx
                    } else if (upOrDown == 66) {
                        ++newCmdIdx
                    }
                    if (newCmdIdx >= 0 && newCmdIdx <= recentCommands.size && newCmdIdx != cmdIdx) {
                        cmdIdx = newCmdIdx
                        if (s.isNotEmpty()) {
                            interop.backspace(s.length)
                            s.delete(0, s.length)
                        }
                        if (newCmdIdx < recentCommands.size) {
                            s.append(recentCommands[cmdIdx])
                        }
                        interop.print(s)
                    }
                }
                else ->
                    if (isPrintableChar(ch)) {
                        if (!isConsole) {
                            interop.printChar(ch)
                        }
                        s.append(ch)
                    }
            }
        }
    }

    private fun toPrintStream(out: OutputStream?): PrintStream? {
        return if (out == null) null else PrintStream(out)
    }

    private val isConsole = true == config[CONSOLE]
}